cellmlmanip/_singularity_fixes.py

```diff
@@ -0,0 +1,252 @@
+"""
+from math import isclose
+
+import networkx as nx
+from sympy import (
+    Abs,
+    Add,
+    Eq,
+    Float,
+    I,
+    Mul,
+    Piecewise,
+    Pow,
+    Wild,
+    exp,
+    log,
+    solveset,
+)
+
+import cellmlmanip
+from cellmlmanip.model import FLOAT_PRECISION
+
+
+ONE = cellmlmanip.Quantity(1.0, 'dimensionless')
+
+
+def _generate_piecewise(expr, V, sp, Vmin, Vmax):
+    """
+    Generates a new (piecewise) expression based on expr with a linear interpolation when V is between Vmin and Vmax
+
+    The returned expression between Vmin and Vmax is::
+    f(vs) + (f(ve) - f(vs)) * (V - vs) / (ve - vs)
+
+    :param: The expression to turn into a piecewise
+    :param: V the voltage variable.
+    :param: the value of the singularity point: the value of V for whih U == 0
+    :param: the value of the lower bound: the value of V for which U - U_offset == 0
+    :param: the value of the upper bound: the value of V for which U + U_offset == 0
+    :return: expr with a singularity exception around sp, if sp is not None else return the original expr
+    """
+    f_Vmin = expr.xreplace({V: Vmin})
+    f_Vmax = expr.xreplace({V: Vmax})
+    return Piecewise((f_Vmin + ((V - Vmin) / (Vmax - Vmin)) * (f_Vmax - f_Vmin),
+                      Abs(V - sp) < Abs((Vmax - Vmin) / 2)), (expr, True))
+
+
+def _get_singularity(expr, V, U_offset, exp_function):
+    """
+    Finds singularities in equations of the form:
+       - ``U / (exp(U) - 1.0)``
+       - ``U / (1.0 - exp(U))``
+       - ``(exp(U) - 1.0) / U``
+       - ``(1.0 - exp(U)) / U``
+
+    In addition to the singularity itself, this functions returns a lower and upper bounds as follows:
+    - sp is the singularity point: the value of V for whih U == 0
+    - Vmin is the value of V for which U - U_offset == 0
+    - Vmax is the value of V for which U + U_offset == 0:
+    :return: (Vmin, Vmax, singularity point)
+    """
+    Z = Wild('Z', real=True)
+    U_wildcard = Wild('U_wildcard', real=True, include=[V])
+    SP_wildcard = Wild('SP_wildcard', real=True)
+
+    def check_bottom_match(m):
+        return m is not None and U_wildcard in m and Z in m and Z != 0
+
+    def check_top_match(m, sp):
+        return m is not None and Z in m and Z != 0 and (sp == m[SP_wildcard]
+                                                        or (isinstance(sp, Float)
+                                                            and isinstance(m[SP_wildcard], Float)
+                                                            and isclose(m[SP_wildcard], sp)))
+
+    def float_dummies(expr):
+        """Turns floats back into Quantity dummies to be in line with the rest of the sympy equations"""
+        if expr is None:
+            return None
+        return expr.xreplace({f: cellmlmanip.Quantity(f, 'dimensionless') for f in expr.atoms(Float)})
+
+    # the denominator is all args where a **-1
+    numerator = tuple(a for a in expr.args if not isinstance(a, Pow) or a.args[1] != -1.0)
```

Comment:
Is this guaranteed to find matching numerator/denominator pairs? Should there be a test to check that?
Maybe this would be better/safer as a single loop that finds fractions, and then make e.g. a list [(num, denom), (num, denom), ...] ?

Reply:
Fair point. I have updated it to include all negative powers (so that for example 1/V**2 which is V**(-2) would be found too. I've also added on the full numerator / denominator to allow the matching to match against the full expression rather than only a part